File: app/src/UI/Features/Records/Activity/forms/common/validators/checkSum.ts
const checkSum = (entries: any[], expected: number, options?: { key: string; readable: string }): boolean | string => {
  const total = entries.reduce((sum, curr) => {
    const rawValue = options?.key ? curr?.[options.key] : curr;
    return sum + (Number(rawValue) || 0);
  }, 0);

  if (total === expected) return true;
  if (options) return `Sum of ${options.readable} must equal ${expected} (current: ${total})`;
  return `Sum must equal ${expected} (current: ${total})`;
};

export default checkSum;

File: api-rework/invasives/api/protocol/activity/validators/check_sum.py
from typing import Any, Iterable, Optional, Union


def check_sum(
    entries: Iterable[Any],
    expected: Union[int, float],
    key: Optional[str] = None,
    readable: Optional[str] = None,
) -> Iterable[Any]:
    """
    Generic logic to ensure the sum of a list (or a specific key within
    a list of objects) matches an expected total.
    """
    total = 0
    for entry in entries:
        raw_value = getattr(entry, key, entry) if key else entry
        try:
            total += float(raw_value or 0)
        except (ValueError, TypeError):
            total += 0

    if total == expected:
        return entries

    label = f"Sum of {readable}" if readable else "Sum"
    raise ValueError(f"{label} must equal {expected} (current: {total})")
